metals/src/main/scala/scala/meta/internal/metals/codeactions/UseNamedArguments.scala

```diff
@@ -0,0 +1,127 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.{Future, ExecutionContext}
+import scala.meta._
+import scala.meta.pc.CancelToken
+import scala.meta.internal.metals._
+import scala.meta.internal.metals.MetalsEnrichments._
+import org.eclipse.{lsp4j => l}
+
+class UseNamedArguments(
+    compilers: Compilers,
+    interactiveSemanticdbs: InteractiveSemanticdbs,
+    trees: Trees
+) extends CodeAction {
+
+  override def kind: String = l.CodeActionKind.Refactor
+
+  override def contribute(
+      params: l.CodeActionParams,
+      token: CancelToken
+  )(implicit ec: ExecutionContext): Future[Seq[l.CodeAction]] = {
+
+    def findMethodApplyOrCtorTreeUnderCursor(
+        root: Tree,
+        cursorPos: l.Range
+    ): Option[Tree] =
+      root
+        .collect {
+          case t: Term.Apply
+              if t.pos.toLSP.encloses(cursorPos) && !t.fun.pos.toLSP.encloses(
+                cursorPos
+              ) =>
+            t
+          case t: Init
+              if t.pos.toLSP.encloses(cursorPos) && !t.tpe.pos.toLSP.encloses(
+                cursorPos
+              ) =>
+            t
+        }
+        .sortBy(_.pos.start)
+        .lastOption
+
+    def buildEdits(
+        tree: Tree,
+        paramNames: List[String]
+    ): List[l.TextEdit] = {
+      val args = tree match {
+        case Term.Apply(_, xs) => xs
+        case Init(_, _, xss) => xss.flatten
+        case _ => Nil
+      }
+      args.zip(paramNames).flatMap {
+        case (Term.Assign(_, _), _) =>
+          // already a named argument, no edit needed
+          None
+        case (term, paramName) =>
+          val position = term.pos.toLSP.getStart
+          val text = s"$paramName = "
+          val edit = new l.TextEdit(new l.Range(position, position), text)
+          Some(edit)
+      }
+    }
+
+    val path = params.getTextDocument().getUri().toAbsolutePath
+
+    val textDocPosParams = new l.TextDocumentPositionParams(
+      params.getTextDocument,
+      params.getRange.getStart
+    )
+
+    val parameterNames: Future[List[String]] = compilers
+      .signatureHelp(textDocPosParams, token, interactiveSemanticdbs)
```

Comment:
Is there a way to distinguish between code actions that are explicitly requested by the users and code action requests that are automatically triggered by the editor?

I am concerned about triggering this method on every `codeAction` request since VS Code requests for code actions on every cursor move. The signature help method triggers typechecking on the same compiler thread as other completion/signatureHelp/hover requests from the user. I'm not sure how well this approach will scale as we introduce more code actions 🤔

The "import missing symbol" code action doesn't suffer from the same problems because it only does expensive analysis when the cursor is over a "no such symbol" diagnostic.

Reply:
to clarify my intention when bringing up `signatureHelp`: I wouldn't merge it as is. I would rather extract the relevant bits off the `SignatureHelpProvider` and provide a more general API to get the active signature info (`signatureHelp` does a lot more, which we don't need for this code action and it also returns data in an inconvenient format).

Reply:
>Is there a way to distinguish between code actions that are explicitly requested by the users and code action requests that are automatically triggered by the editor?

Unfortunately not. On cursor move VS Code sends a code actions request with `only = null`, i.e. it requests all possible actions. Then if the user explicitly requests "Refactor..." or "Quick Fix...", it reuses the info it has already retrieved. I don't think it ever sends a code actions request in which it specifies a value for `only`.

I'm not sure how much work it would be, but as a long term solution would it make sense for `Compilers` (or some layer on top of that) to cache the typechecking results? It could be computed once per `didChange` and then reused by various code actions, or whoever else might need it.

@olafurpg What if we reverted to the previous implementation, using `definitionProvider` and `symbolSearch` instead of the presentation compiler? The code was more cumbersome, but would that alleviate your performance concerns? It was traversing a pre-built scalameta tree, then looking up the symbol in semanticdb, and finally looking up the symbol's definition in the mtags index, all of which should be faster than typechecking?

@gabro That sounds reasonable but IIUC, even if we made the API nicer, any implementation that uses the presentation compiler hits the problem of triggering an expensive typecheck every time we generate the code action.

Not quite sure how to proceed from here...

Reply:
@gabro suggested a nice solution in an offline conversation that the code action only needs to analyze whether the refactoring can be applied at the given position. If the refactoring is eligible, we return an ID for a command that performs the expensive analysis only if user explicitly chooses to apply the code action.

I have a feeling it will be best add one method to the `PresentationCompiler` API that returns what code actions are available at a given position. This method will need to be aware of all available code actions. For "add named argument", this analysis involves checking only if the cursor is above an argument to a method call.

What do you think?

Reply:
By the way, I've discovered that code actions that return a `codeEdit` directly are not necessarily supported by all clients, since they were added in a later revision of LSP. We should honor the `codeActionLiteralSupport` sent by the client.

At a first glance, all major LSP clients seem to support this (I've checked VS Code, coc.nvim, sublime's LSP, emacs-lsp), but we should honor the setting anyway.

I can take care of it in a separate PR.

---

For this specific PR, I have a proposal: can we split it into a PR that contains the refactoring to the codebase and a PR that implements the specific code action?
I think the refactoring is pretty uncontroversial and we could merge it as is, whereas the code action requires a bit more work to nail the PC integration.

What do you think @cb372?

Also, thank you for your patience: we're figuring this out as we go 😅